typing/jkind.ml

```diff
@@ -2608,70 +2645,97 @@ let sub_or_error ~type_equal ~jkind_of_type t1 t2 =
-  (* CR layouts v2.8: Do something better than just comparing for equality. *)
-  (* We can't use other functions, because they insist that we only compare
-     lr-jkinds for equality, not just l-jkinds. *)
-  let layouts =
-    Misc.Le_result.is_le (Layout.sub sub.jkind.layout super.jkind.layout)
+  (* This function implements the "SUB" judgement from kind-inference.md. *)
+  let open Misc.Stdlib.Monad.Result.Syntax in
+  let require_le le_result =
+    match Misc.Le_result.is_le le_result with
+    | true -> Ok ()
+    | false ->
+      (* When we report an error, we want to show the best-normalized version of sub, but
+         the original super. When this check fails, it is usually the case that the super
+         was written by the user and the sub was inferred. Thus, we should display the
+         user-written jkind, but simplify the inferred one, since the inferred one is
+         probably overly complex. *)
+      (* CR layouts v2.8: It would be useful report to the user why this
+         violation occurred, specifically which axes the violation is along. *)
+      let best_sub = normalize ~mode:Require_best ~jkind_of_type sub in
+      Error (Violation.of_ (Not_a_subjkind (best_sub, super)))
+  in
+  let* () =
+    (* Validate layouts *)
+    require_le (Layout.sub sub.jkind.layout super.jkind.layout)
+  in
+  let best_super =
+    (* MB_EXPAND_R *)
+    normalize ~mode:Require_best ~jkind_of_type super
+  in
+  let right_bounds =
+    Type_map.of_best_effort_type_map
+      (With_bounds.to_be_map best_super.jkind.with_bounds)
   in
-  let sub = normalize ~require_best:true ~jkind_of_type sub in
-  (* CR aspsmith: this weird double-normalization is poorly explained because it will go
-     away (soon) once we get better subsumption *)
-  let sub_not_best = lazy (normalize ~require_best:false ~jkind_of_type sub) in
-  let super =
-    normalize ~require_best:true ~jkind_of_type (super |> disallow_right)
+  let irrelevant_axes =
+    (* If the upper_bound is max on the right, then that axis is irrelevant - the
+       left will always satisfy the right along that axis. This is an optimization,
+       not necessary for correctness *)
+    Bounds.get_max_axes best_super.jkind.mod_bounds
   in
-  let bounds =
-    List.for_all
-      (fun (Axis.Pack axis) ->
-        let (module Bound_ops) = Axis.get axis in
-        let bound1 = Bounds.get ~axis sub.jkind.mod_bounds in
-        let bound2 = Bounds.get ~axis super.jkind.mod_bounds in
-        let with_bounds1 =
-          With_bounds.types_on_axis ~axis sub.jkind.with_bounds
+  let ( ({ layout = _;
+           mod_bounds = sub_upper_bounds;
+           with_bounds = No_with_bounds
+         } :
+          Allowance.right_only jkind_desc),
+        _ ) =
+    (* CR layouts v2.8: don't expand along bounds where the upper bound is max on the
+       right. Normalization *)
+    (* MB_EXPAND_L *)
+    Jkind_desc.map_normalize ~jkind_of_type ~mode:Ignore_best
```

Comment:
This step could use a little commentary. I don't quite see what it's doing.

Reply:
I'll add a comment. But because I'm in a hurry right now, a brief explanation: This does the incremental stepping. The left is `Ignore_best` normalized, but before each type is expanded, we look for occurrences of it on the right. If it does occur on the right, then we skip over it.

To go into some more detail on "skipping over": We only skip over axes that are "relevant" to the rhs. ie, if `@@ portable` is on the type on the rhs, we don't skip it. This is why we take the diff of the axes relevant to the left - the axes relevant to the right. And if the resulting set is empty, `map_normalize` is smart enough to not bother expanding the type.

Reply:
I've added a comment that expands on this